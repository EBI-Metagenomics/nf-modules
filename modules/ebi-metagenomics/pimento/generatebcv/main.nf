
process PIMENTO_GENERATEBCV {
    tag "$meta.id"
    label 'process_single'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/mi-pimento:1.0.2--pyhdfd78af_0':
        'biocontainers/mi-pimento:1.0.2--pyhdfd78af_0' }"

    input:
    tuple val(meta), val(fwd_flag), val(rev_flag), path(fastq)

    output:
    tuple val(meta), path("*.tsv"), emit: tsv
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def var_region = "${meta.var_region}" ?: ""
    def assess_mcp_prop_prefix = "${prefix}_${var_region}"
    def strands = ""

    if (fwd_flag == "auto" && rev_flag == "auto") {
        strands = "FR"
    } else if (fwd_flag == "auto") {
        strands = "F"
    } else if (rev_flag == "auto") {
        strands = "R"
    }

    if (strands == "") {
        """
        touch ${assess_mcp_prop_prefix}_mcp_cons.tsv

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            mi-pimento: \$( pimento --version | cut -d" " -f3 )
        END_VERSIONS
        """
    } else {
        """
        pimento \\
            gen_bcv \\
            -i ${fastq} \\
            -st ${strands} \\
            -o ${assess_mcp_prop_prefix}

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            mi-pimento: \$( pimento --version | cut -d" " -f3 )
        END_VERSIONS
        """
    }

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def var_region = "${meta.var_region}"
    def assess_mcp_prop_prefix = "${prefix}_${var_region}"

    """
    touch ${assess_mcp_prop_prefix}_mcp_cons.tsv

    echo 'dummy' > ${assess_mcp_prop_prefix}_mcp_cons.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mi-pimento: \$( pimento --version | cut -d" " -f3 )
    END_VERSIONS
    """
}
