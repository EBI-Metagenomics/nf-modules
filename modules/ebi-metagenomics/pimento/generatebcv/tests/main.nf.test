
nextflow_process {

    name "Test Process PIMENTO_GENERATEBCV"
    script "../main.nf"
    process "PIMENTO_GENERATEBCV"

    tag "modules"
    tag "modules_ebimetagenomics"
    tag "pimento"
    tag "pimento/generatebcv"

    test("pimento/generatebcv - FR") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    "auto",
                    "auto",
                    file("${moduleDir}/tests/fixtures/test.fastq.gz", checkIfExists: true)
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }

    }

    test("pimento/generatebcv - F") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    "auto",
                    "none",
                    file("${moduleDir}/tests/fixtures/test.fastq.gz", checkIfExists: true)
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }

    }

    test("pimento/generatebcv - R") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    "none",
                    "auto",
                    file("${moduleDir}/tests/fixtures/test.fastq.gz", checkIfExists: true)
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }

    }

    test("pimento/generatebcv - none") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    "none",
                    "none",
                    file("${moduleDir}/tests/fixtures/test.fastq.gz", checkIfExists: true)
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }

    }

    test("pimento/generatebcv - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    "auto",
                    "auto",
                    file("${moduleDir}/tests/fixtures/test.fastq.gz", checkIfExists: true)
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.versions.collect { path(it).yaml }
                    ).match()
                }
            )
        }

    }

}
