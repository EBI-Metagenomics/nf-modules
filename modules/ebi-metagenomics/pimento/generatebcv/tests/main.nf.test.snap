{
    "pimento/generatebcv - stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_mcp_cons.tsv:md5,f02e326f800ee26f04df7961adbf7c0a"
                ]
            ],
            [
                {
                    "PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-06-18T14:38:12.95981"
    },
    "pimento/generatebcv - none": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_mcp_cons.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                {
                    "PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-06-18T14:38:04.488138"
    },
    "pimento/generatebcv - F": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_bcv.tsv:md5,1c566bc27d2bf777e76ffb8e52db7132"
                ]
            ],
            [
                {
                    "PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-06-18T14:37:46.746505"
    },
    "pimento/generatebcv - FR": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_bcv.tsv:md5,875c712aadf9b21bd535e555dbb4b7e1"
                ]
            ],
            [
                {
                    "PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-06-18T14:37:37.09592"
    },
    "pimento/generatebcv - R": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_bcv.tsv:md5,e4500d4e1aabab1e143063f4c26029ea"
                ]
            ],
            [
                {
                    "PIMENTO_GENERATEBCV": {
                        "mi-pimento": "1.0.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-06-18T14:37:55.942524"
    }
}